openstudiocore/src/energyplus/ReverseTranslator/ReverseTranslateEnergyManagementSystemProgram.cpp

```diff
@@ -0,0 +1,102 @@
+ **********************************************************************/
+
+#include "../ReverseTranslator.hpp"
+
+#include "../../model/EnergyManagementSystemProgram.hpp"
+#include "../../model/EnergyManagementSystemProgram_Impl.hpp"
+
+#include "../../utilities/idf/IdfExtensibleGroup.hpp"
+#include "../../utilities/idf/WorkspaceExtensibleGroup.hpp"
+#include "../../utilities/core/StringHelpers.hpp"
+
+#include <utilities/idd/EnergyManagementSystem_Program_FieldEnums.hxx>
+#include "../../utilities/idd/IddEnums.hpp"
+#include <utilities/idd/IddEnums.hxx>
+
+using namespace openstudio::model;
+
+namespace openstudio {
+
+namespace energyplus {
+
+OptionalModelObject ReverseTranslator::translateEnergyManagementSystemProgram(const WorkspaceObject & workspaceObject)
+{
+  if (workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_Program) {
+    LOG(Error, "WorkspaceObject is not IddObjectType: EnergyManagementSystem_Program");
+    return boost::none;
+  }
+
+  OptionalString s = workspaceObject.getString(EnergyManagementSystem_ProgramFields::Name);
+  if(!s){
+    LOG(Error, "WorkspaceObject EnergyManagementSystem_Program has no Name");
+    return boost::none;
+  }
+
+  //make sure all other objects are translated first
+  for (const WorkspaceObject& workspaceObject : m_workspace.objects()) {
+    if ((workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_Program) 
+      && (workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_Subroutine)
+      && (workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_ProgramCallingManager)
+      && (workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_MeteredOutputVariable)
+      && (workspaceObject.iddObject().type() != IddObjectType::EnergyManagementSystem_OutputVariable)) {
+      translateAndMapWorkspaceObject(workspaceObject);
+    }
+  }
+
+  openstudio::model::EnergyManagementSystemProgram emsProgram(m_model);
+  emsProgram.setName(*s);
+
+  //get all model objects so we can do name / uid substitution
+  const std::vector<model::ModelObject> modelObjects = m_model.getModelObjects<model::ModelObject>();
+  size_t pos, len;
+  std::string newline, uid;
+
+  unsigned n = workspaceObject.numExtensibleGroups();
+  OptionalString line;
+  for (unsigned i = 0; i < n; ++i) {
+    line = workspaceObject.getExtensibleGroup(i).cast<WorkspaceExtensibleGroup>().getString(EnergyManagementSystem_ProgramExtensibleFields::ProgramLine);
+    if (line) {
+      newline = line.get();
+      //split line on whitespaces to get look for modelobject names
+      std::vector<std::string> results = splitString(line.get(), ' ');
+      for (size_t j = 0; j < results.size(); j++) {
+        for (size_t k = 0; k < modelObjects.size(); k++) {
+          if (modelObjects.at(k).name()) {
+            //check if program item is the name of a model object
+            if (modelObjects.at(k).name().get() == results.at(j)) {
```

Comment:
I'm fairly certain the rest of OS, and auto naming, does case insensitive comparisons for names. This is more strict so it doesn't hurt anything, but you potentially will miss valid objects. @kbenne